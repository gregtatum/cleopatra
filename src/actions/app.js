/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import {
  getAreTooltipsEnabled,
  getIsTooltipDismissRequested,
} from '../reducers/app';
import { getSelectedTab, getDataSource } from '../reducers/url-state';
import { sendAnalytics } from '../utils/analytics';
import type { Action, ThunkAction } from '../types/store';
import type { TabSlug } from '../app-logic/tabs-handling';
import type { ProfileSharingStatus, UrlState } from '../types/reducers';
import type { TooltipReference, MousePosition } from '../types/actions';
import type { CssPixels } from '../types/units';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function setProfileSharingStatus(
  profileSharingStatus: ProfileSharingStatus
): Action {
  return {
    type: 'SET_PROFILE_SHARING_STATUS',
    profileSharingStatus,
  };
}

export function changeTabOrder(tabOrder: number[]): Action {
  return {
    type: 'CHANGE_TAB_ORDER',
    tabOrder,
  };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

/**
 * Tooltips are created from global Redux state. This action causes one to show
 * up in the UI. The action is ignored if tooltips are currently disabled.
 */
export function viewTooltip(
  mouseX: CssPixels,
  mouseY: CssPixels,
  tooltipReference: TooltipReference
): ThunkAction<void> {
  return (dispatch, getState) => {
    if (getAreTooltipsEnabled(getState())) {
      dispatch({
        type: 'VIEW_TOOLTIP',
        tooltipReference,
        mouse: { mouseX, mouseY },
      });
    }
  };
}

/**
 * Dismiss the tooltip immediately.
 */
export function dismissTooltip(): Action {
  return {
    type: 'DISMISS_TOOLTIP',
  };
}

/**
 * Re-position the tooltip.
 */
export function moveTooltip(mouse: MousePosition): Action {
  return {
    type: 'MOVE_TOOLTIP',
    mouse,
  };
}

/**
 * Temporarily disable tooltips from showing. This is useful for stopping tooltips
 * from interfering with mouse actions like draggin.
 */
export function disableTooltips(): Action {
  return {
    type: 'DISABLE_TOOLTIPS',
  };
}

/**
 * Re-enable a tooltip after disabling it.
 */
export function enableTooltips(): Action {
  return {
    type: 'ENABLE_TOOLTIPS',
  };
}

/**
 * It is difficult to coordinate the closing of tooltips between separate components.
 * One may fire the mouseleave, and one will fire the mouseenter. Defer the closing
 * of the tooltip, and initially only request it to be dismissed.
 */
export function requestToDismissTooltip(): ThunkAction<void> {
  return (dispatch, getState) => {
    const { generation: oldGeneration } = getIsTooltipDismissRequested(
      getState()
    );
    const generation = oldGeneration + 1;
    dispatch({
      type: 'REQUEST_TO_DISMISS_TOOLTIP',
      generation,
    });

    // Wait until the next animation frame to allow all of the current event listeners
    // to fire and decide what they want to do.
    requestAnimationFrame(() => {
      const {
        isRequested,
        generation: currentGeneration,
      } = getIsTooltipDismissRequested(getState());

      // Check that the request to
      if (isRequested && generation === currentGeneration) {
        dispatch(dismissTooltip());
      }
    });
  };
}

/**
 * It is difficult to coordinate the closing of tooltips between separate components.
 * One may fire the mouseleave, and one will fire the mouseenter. This action stops
 * the request to dismiss a tooltip, and leaves it open.
 */
export function keepTooltipOpen(): Action {
  return {
    type: 'KEEP_TOOLTIP_OPEN',
  };
}
