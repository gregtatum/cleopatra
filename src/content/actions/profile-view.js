/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  Action, ThunkAction, ProfileSelection, CallTreeFilter, ImplementationFilter,
} from './types';
import type { Thread, ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../../common/types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'HIDE_THREAD',
    threadIndex,
  };
}

export function showThread(threads: Thread[], threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threads,
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}

export function mergeFunction(funcIndex: IndexIntoFuncTable, threadIndex: ThreadIndex) {
  return {
    type: 'MERGE_FUNCTION',
    funcIndex,
    threadIndex,
  };
}

export function unmergeFunction(funcIndex: IndexIntoFuncTable, threadIndex: ThreadIndex) {
  return {
    type: 'UNMERGE_FUNCTION',
    funcIndex,
    threadIndex,
  };
}

export function mergeSubtree(funcIndex: IndexIntoFuncTable, threadIndex: ThreadIndex) {
  return {
    type: 'MERGE_SUBTREE',
    funcIndex,
    threadIndex,
  };
}

export function unmergeSubtree(funcIndex: IndexIntoFuncTable, threadIndex: ThreadIndex) {
  return {
    type: 'UNMERGE_SUBTREE',
    funcIndex,
    threadIndex,
  };
}
