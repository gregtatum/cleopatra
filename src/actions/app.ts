/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { getSelectedTab, getDataSource } from '../selectors/url-state';
import { getTrackThreadHeights } from '../selectors/app';
import { sendAnalytics } from '../utils/analytics';
import { stateFromLocation } from '../app-logic/url-handling';
import { finalizeProfileView } from './receive-profile';
import { Profile, ThreadIndex } from '../types/profile';
import { CssPixels } from '../types/units';
import { Action, ThunkAction } from '../types/store';
import { TabSlug } from '../app-logic/tabs-handling';
import { UrlState } from '../types/state';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function startFetchingProfiles(): Action {
  return { type: 'START_FETCHING_PROFILES' };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when we start setting up the initial url state.
 * It takes the location and profile data, converts the location into url
 * state and then dispatches relevant actions to finalize the view.
 */
export function setupInitialUrlState(
  location: Location,
  profile: Profile
): ThunkAction<void> {
  return dispatch => {
    let urlState;
    try {
      urlState = stateFromLocation(location, profile);
    } catch (e) {
      // The location could not be parsed, show a 404 instead.
      console.error(e);
      dispatch(show404(location.pathname + location.search));
      return;
    }

    // Validate the initial URL state. We can't refresh on a from-file URL.
    if (urlState.dataSource === 'from-file') {
      urlState = null;
    }

    // Normally having multiple dispatches is an anti pattern, but here it's
    // necessary because we are doing different things inside those actions and
    // they can't be merged because we are also calling those seperately on
    // other parts of the code.
    // The first dispatch here updates the url state, then changes state as the url
    // setup is done, and lastly finalizes the profile view since everything is set up now.
    dispatch(updateUrlState(urlState));
    dispatch(urlSetupDone());
    dispatch(finalizeProfileView());
  };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export function reportTrackThreadHeight(
  threadIndex: ThreadIndex,
  height: CssPixels
): ThunkAction<void> {
  return (dispatch, getState) => {
    const trackThreadHeights = getTrackThreadHeights(getState());
    const previousHeight = trackThreadHeights[threadIndex];
    if (previousHeight !== height) {
      // Guard against unnecessary dispatches. This could happen frequently.
      dispatch({
        type: 'UPDATE_TRACK_THREAD_HEIGHT',
        height,
        threadIndex,
      });
    }
  };
}

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}

/**
 * Called when a user has started dragging a file. Used for loading
 * profiles with the drag and drop component.
 */
export function startDragging(): Action {
  return { type: 'START_DRAGGING' };
}

/**
 * Called when a user has stopped dragging a file.
 */
export function stopDragging(): Action {
  return { type: 'STOP_DRAGGING' };
}

/**
 * Called when a custom drag and drop overlay is mounted. This lets
 * the app know that we shouldn't create a default overlay.
 */
export function registerDragAndDropOverlay(): Action {
  return { type: 'REGISTER_DRAG_AND_DROP_OVERLAY' };
}

/**
 * Called when a custom drag and drop overlay is unmounted.
 */
export function unregisterDragAndDropOverlay(): Action {
  return { type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY' };
}
