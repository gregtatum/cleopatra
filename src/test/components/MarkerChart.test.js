/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import { changeMarkersSearchString } from '../../actions/profile-view';
import MarkerChart from '../../components/marker-chart';
import { changeSelectedTab } from '../../actions/app';

import EmptyReasons from '../../components/shared/EmptyReasons';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithMarkers } from '../fixtures/profiles/make-profile';
import { getBoundingBox } from '../fixtures/utils';
import mockRaf from '../fixtures/mocks/request-animation-frame';

const MARKERS = [
  ['Marker A', 0, { startTime: 0, endTime: 10 }],
  ['Marker A', 0, { startTime: 0, endTime: 10 }],
  ['Marker A', 11, { startTime: 11, endTime: 15 }],
  [
    'Very very very very very very Very very very very very very Very very very very very very Very very very very very very Very very very very very very long Marker D',
    6,
    { startTime: 5, endTime: 15 },
  ],
  ['Dot marker E', 4, { startTime: 4, endTime: 4 }],
  ['Non-interval marker F without data', 7, null],
  [
    'Marker G type DOMEvent',
    5,
    {
      type: 'tracing',
      category: 'DOMEvent',
      eventType: 'click',
      interval: 'start',
      phase: 2,
    },
  ],
  [
    'Marker G type DOMEvent',
    10,
    {
      type: 'tracing',
      category: 'DOMEvent',
      eventType: 'click',
      interval: 'end',
      phase: 2,
    },
  ],
];

function setupWithProfile(profile) {
  const flushRafCalls = mockRaf();
  const ctx = mockCanvasContext();
  jest
    .spyOn(HTMLCanvasElement.prototype, 'getContext')
    .mockImplementation(() => ctx);

  // Ideally we'd want this only on the Canvas and on ChartViewport, but this is
  // a lot easier to mock this everywhere.
  jest
    .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
    .mockImplementation(() => getBoundingBox(200, 300));

  const store = storeWithProfile(profile);
  store.dispatch(changeSelectedTab('marker-chart'));

  const markerChart = mount(
    <Provider store={store}>
      <MarkerChart />
    </Provider>
  );

  return {
    markerChart,
    flushRafCalls,
    dispatch: store.dispatch,
    flushDrawLog: () => ctx.__flushDrawLog(),
  };
}

describe('MarkerChart', function() {
  it('renders the normal marker chart and matches the snapshot', () => {
    window.devicePixelRatio = 1;

    const profile = getProfileWithMarkers([...MARKERS]);
    const {
      flushRafCalls,
      dispatch,
      markerChart,
      flushDrawLog,
    } = setupWithProfile(profile);

    dispatch(changeSelectedTab('marker-chart'));
    markerChart.update();
    flushRafCalls();

    const drawCalls = flushDrawLog();
    expect(markerChart).toMatchSnapshot();
    expect(drawCalls).toMatchSnapshot();

    delete window.devicePixelRatio;
  });

  it('renders the hoveredItem markers properly', () => {
    window.devicePixelRatio = 1;

    const profile = getProfileWithMarkers(MARKERS);
    const {
      flushRafCalls,
      dispatch,
      markerChart,
      flushDrawLog,
    } = setupWithProfile(profile);

    dispatch(changeSelectedTab('marker-chart'));
    markerChart.update();
    flushRafCalls();
    flushDrawLog();

    // No tooltip displayed yet
    expect(markerChart.find('Tooltip').exists()).toEqual(false);

    // Move the mouse on top of an item.
    markerChart.find('canvas').simulate('mousemove', {
      nativeEvent: { offsetX: 50, offsetY: 5 },
      pageX: 50,
      pageY: 5,
    });
    markerChart.update();
    flushRafCalls();

    const drawCalls = flushDrawLog();
    expect(drawCalls).toMatchSnapshot();

    // The tooltip should be displayed
    expect(markerChart.find('Tooltip').exists()).toEqual(true);
  });

  describe('with search strings', function() {
    function getFillTextCalls(drawCalls) {
      return drawCalls
        .filter(([methodName]) => methodName === 'fillText')
        .map(([_, text]) => text);
    }

    const searchString = 'Dot marker E';

    it('renders lots of markers initially', function() {
      const profile = getProfileWithMarkers(MARKERS);
      const { flushRafCalls, flushDrawLog } = setupWithProfile(profile);

      flushRafCalls();
      const text = getFillTextCalls(flushDrawLog());
      expect(text.length).toBeGreaterThan(1);
      // Check that our test search string is in here:
      expect(text.filter(t => t === searchString).length).toBe(1);
    });

    it('renders only the marker that was searched for', function() {
      const profile = getProfileWithMarkers(MARKERS);
      const {
        flushRafCalls,
        dispatch,
        flushDrawLog,
        markerChart,
      } = setupWithProfile(profile);

      // Flush out any existing draw calls.
      flushRafCalls();
      flushDrawLog();

      // Update the chart with a search string.
      dispatch(changeMarkersSearchString(searchString));
      markerChart.update();
      flushRafCalls();

      const text = getFillTextCalls(flushDrawLog());
      expect(text.length).toBe(1);
      expect(text[0]).toBe(searchString);
    });
  });

  describe('EmptyReasons', () => {
    it('shows a reason when a profile has no markers', () => {
      const profile = getProfileWithMarkers([]);
      const { dispatch, markerChart } = setupWithProfile(profile);

      dispatch(changeSelectedTab('marker-chart'));
      markerChart.update();
      expect(markerChart.find(EmptyReasons)).toMatchSnapshot();
    });

    it("shows a reason when a profile's markers have been filtered out", () => {
      const profile = getProfileWithMarkers(MARKERS);
      const { dispatch, markerChart } = setupWithProfile(profile);

      dispatch(changeSelectedTab('marker-chart'));
      dispatch(changeMarkersSearchString('MATCH_NOTHING'));
      markerChart.update();
      expect(markerChart.find(EmptyReasons)).toMatchSnapshot();
    });
  });
});
