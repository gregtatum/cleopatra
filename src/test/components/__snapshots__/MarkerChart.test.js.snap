// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkerChart EmptyReasons shows a reason when a profile has no markers 1`] = `
<div
  className="EmptyReasons"
>
  <h3>
    The 
    marker chart
     is empty for “
    Empty
    ”
  </h3>
  <p>
    This thread contains no markers for this chart.
  </p>
</div>
`;

exports[`MarkerChart EmptyReasons shows a reason when a profile's markers have been filtered out 1`] = `
<div
  className="EmptyReasons"
>
  <h3>
    The 
    marker chart
     is empty for “
    Empty
    ”
  </h3>
  <p>
    All markers were filtered out by the current selection or search term.
  </p>
</div>
`;

exports[`MarkerChart renders the hoveredItem markers properly 1`] = `
Array [
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    200,
    300,
  ],
  Array [
    "set lineWidth",
    1,
  ],
  Array [
    "set fillStyle",
    "Highlight",
  ],
  Array [
    "fillRect",
    1,
    1,
    248,
    1,
  ],
  Array [
    "fillRect",
    0,
    2,
    250,
    11,
  ],
  Array [
    "fillRect",
    1,
    13,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    1,
    17,
    248,
    1,
  ],
  Array [
    "fillRect",
    0,
    18,
    250,
    11,
  ],
  Array [
    "fillRect",
    1,
    29,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "arc",
    105,
    39.5,
    3.75,
    0,
    6.283185307179586,
  ],
  Array [
    "fill",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    126,
    49,
    248,
    1,
  ],
  Array [
    "fillRect",
    125,
    50,
    250,
    11,
  ],
  Array [
    "fillRect",
    126,
    61,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    126,
    65,
    123,
    1,
  ],
  Array [
    "fillRect",
    125,
    66,
    125,
    11,
  ],
  Array [
    "fillRect",
    126,
    77,
    123,
    1,
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "click",
    128,
    75,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "arc",
    180,
    87.5,
    3.75,
    0,
    6.283185307179586,
  ],
  Array [
    "fill",
  ],
  Array [
    "set fillStyle",
    "#eee",
  ],
  Array [
    "fillRect",
    0,
    15,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    63,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    79,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    95,
    200,
    1,
  ],
  Array [
    "createLinearGradient",
    0,
    0,
    150,
    0,
  ],
  Array [
    "addColorStop",
    0,
    "rgba(255, 255, 255, 0.8)",
  ],
  Array [
    "addColorStop",
    1,
    "rgba(255, 255, 255, 0.0)",
  ],
  Array [
    "set fillStyle",
    Object {
      "addColorStop": [MockFunction] {
        "calls": Array [
          Array [
            0,
            "rgba(255, 255, 255, 0.8)",
          ],
          Array [
            1,
            "rgba(255, 255, 255, 0.0)",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
  ],
  Array [
    "fillRect",
    0,
    0,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    16,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    32,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    48,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    64,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    80,
    150,
    15,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "fillText",
    "Marker A",
    5,
    11,
  ],
  Array [
    "fillText",
    "Dot marker E",
    5,
    43,
  ],
  Array [
    "fillText",
    "Very very very very very very Very very very very very very Very very very very very very Very very very very very very Very very very very very very …",
    5,
    59,
  ],
  Array [
    "fillText",
    "Marker G type DOMEvent",
    5,
    75,
  ],
  Array [
    "fillText",
    "Non-interval marker F without data",
    5,
    91,
  ],
]
`;

exports[`MarkerChart renders the normal marker chart and matches the snapshot 1`] = `
<div
  className="markerChart"
>
  <div
    className="markerSettings"
  >
    <div
      className="markerSettingsSpacer"
    />
    <div
      className="markerSettingsSearchbar"
    >
      <label
        className="markerSettingsSearchbarLabel"
      >
        Filter Markers: 
        <form
          className="idleSearchField markerSettingsSearchField"
          onSubmit={[Function]}
        >
          <input
            className="idleSearchFieldInput"
            name="search"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required="required"
            title="Only display markers that match a certain name"
            type="search"
            value=""
          />
          <input
            className="idleSearchFieldButton"
            onClick={[Function]}
            onFocus={[Function]}
            type="reset"
          />
        </form>
      </label>
    </div>
  </div>
  <div
    className="chartViewport"
    onMouseDown={[Function]}
    onWheel={[Function]}
  >
    <div>
      <canvas
        className="chartCanvas markerChartCanvas"
        onDoubleClick={[Function]}
        onMouseDown={[Function]}
        onMouseMove={[Function]}
        onMouseOut={[Function]}
      />
    </div>
    <div
      className="chartViewportShiftScroll hidden"
    >
      Zoom Chart:
      <kbd
        className="chartViewportShiftScrollKbd"
      >
        Shift
      </kbd>
      <kbd
        className="chartViewportShiftScrollKbd"
      >
        Scroll
      </kbd>
    </div>
  </div>
</div>
`;

exports[`MarkerChart renders the normal marker chart and matches the snapshot 2`] = `
Array [
  Array [
    "scale",
    1,
    1,
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    200,
    300,
  ],
  Array [
    "set lineWidth",
    1,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "measureText",
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.()< /:-_",
  ],
  Array [
    "measureText",
    "…",
  ],
  Array [
    "fillRect",
    1,
    1,
    248,
    1,
  ],
  Array [
    "fillRect",
    0,
    2,
    250,
    11,
  ],
  Array [
    "fillRect",
    1,
    13,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    1,
    17,
    248,
    1,
  ],
  Array [
    "fillRect",
    0,
    18,
    250,
    11,
  ],
  Array [
    "fillRect",
    1,
    29,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "arc",
    105,
    39.5,
    3.75,
    0,
    6.283185307179586,
  ],
  Array [
    "fill",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    126,
    49,
    248,
    1,
  ],
  Array [
    "fillRect",
    125,
    50,
    250,
    11,
  ],
  Array [
    "fillRect",
    126,
    61,
    248,
    1,
  ],
  Array [
    "measureText",
    "",
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    126,
    65,
    123,
    1,
  ],
  Array [
    "fillRect",
    125,
    66,
    125,
    11,
  ],
  Array [
    "fillRect",
    126,
    77,
    123,
    1,
  ],
  Array [
    "measureText",
    "click",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "click",
    128,
    75,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "beginPath",
  ],
  Array [
    "arc",
    180,
    87.5,
    3.75,
    0,
    6.283185307179586,
  ],
  Array [
    "fill",
  ],
  Array [
    "set fillStyle",
    "#eee",
  ],
  Array [
    "fillRect",
    0,
    15,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    63,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    79,
    200,
    1,
  ],
  Array [
    "fillRect",
    0,
    95,
    200,
    1,
  ],
  Array [
    "createLinearGradient",
    0,
    0,
    150,
    0,
  ],
  Array [
    "addColorStop",
    0,
    "rgba(255, 255, 255, 0.8)",
  ],
  Array [
    "addColorStop",
    1,
    "rgba(255, 255, 255, 0.0)",
  ],
  Array [
    "set fillStyle",
    Object {
      "addColorStop": [MockFunction] {
        "calls": Array [
          Array [
            0,
            "rgba(255, 255, 255, 0.8)",
          ],
          Array [
            1,
            "rgba(255, 255, 255, 0.0)",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      },
    },
  ],
  Array [
    "fillRect",
    0,
    0,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    16,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    32,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    48,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    64,
    150,
    15,
  ],
  Array [
    "fillRect",
    0,
    80,
    150,
    15,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "fillText",
    "Marker A",
    5,
    11,
  ],
  Array [
    "fillText",
    "Dot marker E",
    5,
    43,
  ],
  Array [
    "fillText",
    "Very very very very very very Very very very very very very Very very very very very very Very very very very very very Very very very very very very …",
    5,
    59,
  ],
  Array [
    "fillText",
    "Marker G type DOMEvent",
    5,
    75,
  ],
  Array [
    "fillText",
    "Non-interval marker F without data",
    5,
    91,
  ],
]
`;
